src/main/java/com/google/devtools/build/lib/query2/query/output/StreamedProtoOutputFormatter.java

```diff
@@ -34,13 +50,76 @@ public String getName() {
   public OutputFormatterCallback<Target> createPostFactoStreamCallback(
       final OutputStream out, final QueryOptions options, LabelPrinter labelPrinter) {
     return new OutputFormatterCallback<Target>() {
+      private final LabelPrinter ourLabelPrinter = labelPrinter;
+
       @Override
       public void processOutput(Iterable<Target> partialResult)
           throws IOException, InterruptedException {
-        for (Target target : partialResult) {
-          toTargetProtoBuffer(target, labelPrinter).writeDelimitedTo(out);
+        try {
+          StreamSupport.stream(partialResult.spliterator(), /* parallel= */ true)
```

Comment:
I'm not familiar with parallel streams, but I _think_ if this is interrupted then only one of the parallel items sees it and exits, while the rest will carry on. I don't think we want this since it will leave lingering threads doing formatting and writing to the output possibly beyond the output being closed or the command itself having exited. I think encountering an IOException has a similar issue - every thread will keep going and hitting the io exception even after this has exited.

We have a similar parallel formatting implementation for some internal code[^1] - IIRC we use `close` as a synchronization point to make sure that nothing is left behind from `processOutput`, as well as to reconcile any concurrent or cascading exceptions. What you're trying to do might be different enough that you can avoid this, but I fear you also might not get so lucky.

[^1]: A little too tightly coupled with some internal-only code to easily open source, coming from having tried a while ago

Reply:
Ok, worth investigating. I will test ctrl+c'ing out of it. If needed I will make a little test case to determine runtime behavior of interrupts on parallel streams.

Reply:
Ctrl+c does not respond quickly! I will need to find a better solution.

Reply:
Note: ctrl+c responds immediately with the current `ForkJoinPool`-based implementation!